storage/rocksdb/ha_rocksdb.cc

```diff
@@ -10619,29 +10625,37 @@ void Rdb_background_thread::run()
 
     if (local_stop)
     {
+      // If we're here then that's because condition variable was signaled by
+      // another thread and we're shutting down. Break out the loop to make
+      // sure that shutdown thread can proceed.
+      DBUG_ASSERT(ret == 0);
       break;
     }
 
+    // This path should be taken only when the timer expired.
+    DBUG_ASSERT(ret == ETIMEDOUT);
+
     if (local_save_stats)
     {
       ddl_manager.persist_stats();
     }
 
-    // Flush the WAL if need be but don't do it more frequent
-    // than once per second
     timespec ts;
     clock_gettime(CLOCK_REALTIME, &ts);
-    if (ts.tv_sec - ts_next_sync.tv_sec >= 1)
+
+    // Flush the WAL.
+    if (rdb && rocksdb_background_sync)
```

Comment:
The rdb_bg_thread is used to both sync the wal file in the background and flush out the cached stats. rdb_bg_thread.request_save_stats() is called randomly based on compaction, memtable flushes. If it happens frequently, then on each loop, we call SyncWal(). The 1 second SyncWal behavior was meant to emulate Innodb's flush_log_at_trx_commit=2 behavior: http://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_flush_log_at_trx_commit.

While this code is functional and correct, we might end up calling SyncWal() multiple times in a second.

An alternative would be to ignore the signals from rdb_queue_save_stats_request(). In that case, this thread would only wake up once a second to perform syncs and stat flushes.